File: LittleProjects/Java/SpringHibernateCRUD/src/main/java/core/repository/Repository.java
package core.repository;

import java.io.Serializable;

public interface Repository<T, ID extends Serializable> {
    T select(ID id);

    Iterable<T> select();

    void deleteById(ID id);

    void deleteById(Iterable<ID> ids);

    void delete(T entity);

    void delete(Iterable<T> entities);

    T insert(T entity);

    Iterable<T> insert(Iterable<T> entities);

    T update(T entity);

    Iterable<T> update(Iterable<T> entities);
}

File: LittleProjects/Scala/ScalaSpringSimpleCRUD/src/main/java/com/github/gr1f0n6x/repository/Repository.scala
package com.github.gr1f0n6x.repository

import scalikejdbc.AutoSession

trait Repository[T, ID] {
  implicit val session: AutoSession.type = AutoSession

  def select(id: ID): Option[T]

  def select(): Seq[T]

  def insert(entity: T): Unit

  def insert(entities: Seq[T]): Unit

  def delete(id: ID): Unit

  def delete(ids: Seq[ID]): Unit

  def update(entity: T): Unit

  def update(entities: Seq[T]): Unit
}
